### DISCUSSION TITLE
issue: setError stops working if using it with an unexisting field

### DISCUSSION THREAD

**Discussion by newfylox** [react-hook-form/react-hook-form#9906] [Q&A]
### Version Number

7.43.1

### Codesandbox/Expo snack

https://codesandbox.io/s/determined-maria-82xf6k

### Steps to reproduce

1. Click on "submit" button to see that `handleSubmit` is triggering `onSubmit`
1. `onSubmit` will have errors from server, also having custom errors that don't belong to the form
1. Click again to see that it doesn't work anymore because `setError` had set an error on an unexisting field

### Expected behaviour

I'm expecting that even if I set an error on a field that doesn't belong to the form, the `handleSubmit` while continue to behave as normal, but it crashed somewhere.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Please refer to the doc's rule section: https://react-hook-form.com/api/useform/seterror

![Screenshot 2023-02-10 at 9 43 29 am](https://user-images.githubusercontent.com/10513364/217955899-62cabae0-6c42-4529-badf-ba87be4f378d.png)

**Comment by leapful** [comment]
@newfylox 

Please following the documentation of Bill for setting custom error. It should be "root.customServerError". Furthermore, in your example, you are using error on field `firstName` which is an registered field on react-hook-form too therefore end-user needs to solve the error of `firstName` to continue on submitting, in case you want to use it with custom error.

Working demo: https://codesandbox.io/s/react-hook-form-seterror-forked-isji2c?file=/src/App.js

**Reply by newfylox** [reply]
I understand by reading the documentation, but I guess there's something I'm not really getting it.

In my example, it wasn't about a custom server error (sorry I misinformed you) but a validation error. For example, in my API, I added a new validation, therefore a new field to my JSON response. But the problem is that I don't think my backend should be aware of the frontend. My backend should be able to be updated without caring of what the frontend would have. Let's say I added `phone_number` as a new field in my JSON response but my website is only using `first_name` and `last_name`, with my global `setErrors` helper it will crash using `setError` with `phone_number` field.

In the opposite, I don't think I should sanitize every fields every time I'm using my helper in my frontend.

Like I said, probably I don't get it and you would surely be able to help me understand it.

**Reply by leapful** [reply]
@newfylox 

The method `handleSubmit(onValid, onError)` from `useForm()` hook only triggers `onValid` callback if there is no error.

In your example, you function handler named `onSubmit` is `onValid` callback for `handleSubmit`.

You are setting error manually on unregistered field `phone_number` without prefix `root.`, therefore, the error still persists even after submitting because the `onValid` callback never be triggered since manual error `phone_number` exists.

It's not reasonable to have a custom field `phone_number` which does not belong to your whole form but now acts like a registered form that needs to clear error automatically by the form. If react-hook-form does not know about the field name via registration, how can it clear error automatically ? That's why convention adding prefix `root.` is introduced to help react-hook-form clear custom error that is not needed to be persisted.

If you really want to achieve that behavior, just register the `phone_number` with react-hook-form but then exclude on form data when submitting to back-end side.

**Reply by newfylox** [reply]
I understand... so my frontend will always have to be aware of backend updates? I mean if I want to register those new added fields in my JSON response from my API, I will always have to update my frontend to register those new fields?

Thanks again!

**Reply by newfylox** [reply]
I forgot to mention, I'm using `react-hook-form` with `react-native` as described in the docs. So I guess that `just register the field` is when I'm using the `<Controller />` component with the `name` property? If so then you're telling me that I need to register `phone_number` as an input (maybe hidden) and ignore it in the backend?

But still I have the previous concern that I mentioned, having to register fields every time I update my JSON response.
